core/bin/migrations/20240603_backfill_stored_documents.rs

```diff
@@ -0,0 +1,348 @@
+use ::qdrant_client::qdrant::{value, PointId};
+use anyhow::{anyhow, Context, Result};
+use cloud_storage::Object;
+use dust::{
+    data_sources::{
+        data_source::{make_document_id_hash, DataSource, Section},
+        file_storage_document::FileStorageDocument,
+        qdrant::QdrantClients,
+    },
+    stores::{postgres, store},
+    utils,
+};
+use qdrant_client::qdrant;
+use tokio_postgres::Row;
+
+use bb8::Pool;
+use bb8_postgres::PostgresConnectionManager;
+use futures::{StreamExt, TryStreamExt};
+use tokio_postgres::NoTls;
+use tokio_stream::{self as stream};
+
+async fn fetch_data_sources_batch(
+    pool: &Pool<PostgresConnectionManager<NoTls>>,
+    last_id: u64,
+    limit: usize,
+) -> Result<Vec<Row>, anyhow::Error> {
+    let c = pool.get().await?;
+
+    c.query(
+        "SELECT id, internal_id FROM data_sources WHERE id > $1 LIMIT $2",
+        &[&(last_id as i64), &(limit as i64)],
+    )
+    .await
+    .context("Query execution failed")
+}
+
+// Return a "best-effort" section for the latest version of the document (based on Qdrant).
+// For superseded versions, return an empty section.
+async fn get_sections_from_document_version(
+    qdrant_clients: &QdrantClients,
+    ds: &DataSource,
+    document_id_hash: &str,
+    document_status: &str,
+    full_text: &str,
+) -> Result<Section> {
+    if document_status != "latest" {
+        return Ok(Section {
+            prefix: None,
+            content: Some(full_text.to_string()),
+            sections: vec![],
+        });
+    }
+
+    let qdrant_client = ds.main_qdrant_client(qdrant_clients);
+
+    let f = qdrant::Filter {
+        must: vec![qdrant::Condition::matches(
+            "document_id_hash",
+            document_id_hash.to_string(),
+        )],
+        ..Default::default()
+    };
```

Comment:
Can we add just toooo be sure the data source over filter to ensure we never look at someone else's data?

Reply:
Added.